plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/CloudLibraries.java

```diff
@@ -65,28 +72,47 @@
   }
   
   /**
-   * @return the library with the specified ID, or null if not found
+   * Returns the library with the specified ID, or null if not found
    */
   public static Library getLibrary(String id) {
-    // todo this method is only called by LibraryClasspathContainerResolverService and tests.
-    // Perhaps here or in LibraryClasspathContainerResolverService is where we should 
-    // resolve the transitive dependencies
     return libraries.get(id);
   }
+
+  private static final LoadingCache<IJavaProject, Library> masterLibraries =
```

Comment:
Caching `Library` doesn't seem the right direction. The cache is gone when Eclipse restarts, which means that you have no info about which projects have which libraries. Steps to reproduce a problem:

1. Create a fresh project and add Endpoints and Objectify. Verify they are both added.
1. _File_ > _Restart_ (or restart Eclipse by any means)
1. Add App Engine API to the same project.

The project will remove Objects and Endpoints added earlier.

Reply:
Makes sense. Let me think about this. I suspect we need a ClasspathContainerInitializer that preloads the cache. Or some other way of retrieving the container from the IJavaProject.

Reply:
I'm having trouble figuring out how to read the specific jars from the master container. There may be some massive hacks that can make this work, but it feels like there should be a more straightforward way to list out the contents of the master-container once I have a reference to it.

Reply:
The previous approach took the Library ID from the container path.  I think you'll need to explicitly record the selected Library IDs either into the serialized JSON, or in a separate file alongside.

Reply:
Any idea how we might  record the selected Library IDs into the serialized JSON? I can't find any API or documentation for this file. There must be some code in Eclipse somewhere that's writing this stuff, probably in some internal package. 

Alternatively, if there were an API for reading the actual jar paths out of an IClasspathEntry that represents a container, that would let us make this work.

Reply:
It's our responsibility to read/write the jar list.  The Maven container transforms the pom.xml into a list of jars; the PDE container transforms the Require-Bundles and Import-Packages into a list of jars.

So we need a single method, `CloudLibraries.load(IJavaProject, String containerId)`.  If a `Library` exists with ID `containerId`, then return it, otherwise look for a serialized library descriptor for the `IJavaProject` (e.g., in `.settings/com.google.cloud.tools.eclipse.xxx/<containerId>.json`) and create a `Library` from that previously-serialized descriptor.